src/mlpack/tests/rbm_network_test.cpp

```diff
@@ -0,0 +1,172 @@
+#include <mlpack/core.hpp>
+
+#include <mlpack/core/optimizers/rmsprop/rmsprop.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/rbm.hpp>
+#include <mlpack/methods/softmax_regression/softmax_regression.hpp>
+#include <mlpack/core/optimizers/minibatch_sgd/minibatch_sgd.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+#include <mlpack/core/optimizers/lbfgs/lbfgs.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::regression;
+
+BOOST_AUTO_TEST_SUITE(RbmNetworkTest);
+
+template<typename MatType = arma::mat>
+void BuildVanillaNetwork(MatType& trainData,
+                         const size_t hiddenLayerSize)
+{
+  /*
+   * Construct a feed forward network with trainData.n_rows input nodes,
+   * hiddenLayerSize hidden nodes and trainLabels.n_rows output nodes. The
+   * network structure looks like:
+   *
+   *  Visible       Hidden        
+   *  Layer         Layer         
+   * +-----+       +-----+       
+   * |     |       |     |            
+   * |     +<----->|     |
+   * |     |       |     | 
+   * +-----+       +-----+     
+   *        
+   */
+  arma::mat output;
+  BinaryLayer<> visible(trainData.n_rows, hiddenLayerSize, 1);
+  BinaryLayer<> hidden(hiddenLayerSize, trainData.n_rows, 0);
+  GaussianInitialization gaussian(0, 0.1);
+  RBM<GaussianInitialization, BinaryLayer<>, BinaryLayer<>> model(trainData,
+      gaussian, visible, hidden, 1,  true);
+  MiniBatchSGD msgd(10, 0.06, trainData.n_cols * 20, 0, true);
+  model.Train(trainData, msgd);
+  model.Reset();
+  // Set the parmaeters from a learned rbm
+  model.Parameters() = arma::mat(
+      "-0.23224054, -0.23000632, -0.25701271, -0.25122418, -0.20716651"
+      "-0.20962217, -0.59922456, -0.60003836, -0.6, -0.625, -0.475;");
+  // Check Weight Shared
+  BOOST_REQUIRE_CLOSE(arma::accu(model.VisibleLayer().Weight() -
```

Comment:
I think the weights should be equal. Why did you choose 1e-2 accuracy? The precision of `double` around 1 is equal to `1e-16` so you can set something like `1e-14`.